### ISSUE TITLE
Can't push lfs files after updating to 2.5.0

### ISSUE THREAD

**Issue by dnewmarch** [git-lfs/git-lfs#3171]
I'm running git 1.9.1 on Ubuntu 14.04, and I updated my git-lfs from 2.4.2 to 2.5.0 when the update became available recently.

I'm using git-lfs with a locally hosted Bitbucket server running v5.11.1.

After updating my git-lfs package to 2.5.0, when I try to push lfs files to the server the upload bitrate  and progress go a bit nuts, and eventually I get the following error:
Fatal error: Unable to parse HTTP response for PUT https://<servername>/rest/git-lfs/storage/<path to lfs object>: unexpected EOF
error: failed to push some refs to 'ssh://git@<project>.git'

If I downgrade my git-lfs package back to 2.4.2 the problem goes away, but using 2.5.0 or 2.5.1 git-lfs pushing is broken for me.

**Comment by ttaylorr** [comment]
> After updating my git-lfs package to 2.5.0, when I try to push lfs
> files to the server the upload bitrate and progress go a bit nuts,

Oh, no -- I'm sorry that there was a regression and I'm sure that we'll
be able to get to the bottom of whatever is going on.

> and eventually I get the following error: Fatal error: Unable to parse
> HTTP response for PUT https://<servername>/rest/git-lfs/storage/<path
> to lfs object>: unexpected EOF error: failed to push some refs to
> 'ssh://git@<project>.git'

Can you reproduce the issue reliably? If so, please post the output
(with your sensitive credentials obfuscated) by re-running the command
with the following in your environment: `GIT_TRACE=1` and
`GIT_CURL_VERBOSE=1`. That should help us pinpoint a little closer what
the issue might be.

Another thing that you could try is:

    $ git config lfs.contenttype 0

Which will disable "Content-Type:" HTTP header inference.

**Comment by ar613** [comment]
I have the same problem.

2.5.1 seems to send text/plain sometimes.
```
> Content-Type: text/plain; charset=utf-8
> User-Agent: git-lfs/2.5.1 (GitHub; linux amd64; go devel +bd98a81dc2 Thu Jul 26 17:13:12 2018 +0000; git e1628484)
> 
test
13:57:14.402592 trace git-lfs: HTTP: 500
```

Wheras 2.4.1 sent the same file as application/octet-stream
```
> Content-Type: application/octet-stream
> User-Agent: git-lfs/2.4.1 (GitHub; linux amd64; go 1.10; git 79336479)
> 
14:23:13.007511 trace git-lfs: HTTP: 200
```

` git config lfs.contenttype 0` seems to work.
```
> Content-Length: 7
> User-Agent: git-lfs/2.5.1 (GitHub; linux amd64; go devel +bd98a81dc2 Thu Jul 26 17:13:12 2018 +0000; git e1628484)
>
14:31:33.053188 trace git-lfs: HTTP: 200
```

The Bitbucket Server Git LFS REST API handler has @Consumes(MediaType.APPLICATION_OCTET_STREAM)  but Git LFS 2.5.0+ does not always use application/octet-stream

Seems to be a direct result of this commit: https://github.com/git-lfs/git-lfs/commit/231cc3ae51cb7cd6ed72c9437fd462e773946ade

**Comment by mmielke-ciena** [comment]
What is the value of using a detected Content-Type here? Isn't Git LFS supposed to preserve "raw bytes", and the only documentation specifies application/octet-stream:

https://github.com/git-lfs/git-lfs/blob/master/docs/api/basic-transfers.md

My understanding is that any other types, particularly including text/plain, could allow for some ambiguity in "raw bytes", in that "\r\n" vs "\n" might become ambiguously handled by the web server?

Why not keep it simple, and leave it as it has always been here? Did something come up that requires the Content-Type to be preserved?

**Comment by ttaylorr** [comment]
> The Bitbucket Server Git LFS REST API handler has
> @Consumes(MediaType.APPLICATION_OCTET_STREAM)  but Git LFS 2.5.0+ does
> not always use application/octet-stream

I see. It sounds like that is the problem. In this case, you can run:

    $ git config lfs.contenttype false

To disable content-type inference.

I am advocating that we keep the feature since it is disable-able, and
will open the doors for features on the server side (like pluggable
storage backends based on filetype, and etc.)

**Comment by mmielke-ciena** [comment]
The "git config" options is impractical, as we have hundreds of users who don't want to know about Git LFS work-arounds due to unexpected breakage between the client and the server and this means asking each of these users to correctly type that "git config" command (or some other more coordinated way of pushing configuration into their individual workspaces on many different platforms). So, we're left with testing Git LFS before making it available to users.

In this particular case, I was able to modify our proxy between the users and Bitbucket, to override the Content-Type and force it to application/octet-stream for HTTP PUT of this one URL, and this seems to work.

Going forwards, I do think there needs to be more care taken about compatibility. The specification still seems to suggest that only application/octet-stream is valid, and I think Atlassian was correct to implement their code the way they did. That it is now broken will be frustrating to everybody impacted including us and them.

**Comment by ttaylorr** [comment]
> The "git config" options is impractical, as we have hundreds of users
> who don't want to know about Git LFS work-arounds due to unexpected
> breakage between the client and the server and this means asking each
> of these users to correctly type that "git config" command (or some
> other more coordinated way of pushing configuration into their
> individual workspaces on many different platforms).

I can understand your frustration in this regard. We do have support for
a limited set of pre-defined configuration options that you can
distribute to all users of a repository via a specially-named
`.lfsconfig` file in the root of your repository.

It only accepts from a list of known-safe keys, and lfs.contenttype is
not one of them (yet).

The only thing standing in the way of us adding lfs.contenttype is
allowing the optional URL-level configuration such that we would also
accept 'lfs.yourgithost.contenttype', too.

That might not be such an issue, if other members of @git-lfs/core would
be OK with just support lfs.contenttype as a safe key as a punt for more
robust lfs.<url>.contenttype-like keys in the future. I'd love to
collect thoughts from other members of @git-lfs/core on this issue here.

> In this particular case, I was able to modify our proxy between the
> users and Bitbucket, to override the Content-Type and force it to
> application/octet-stream for HTTP PUT of this one URL, and this seems
> to work.

OK, it sounds like you have a work-around in place. I'm sorry that you
had to implement it in this way.

> Going forwards, I do think there needs to be more care taken about
> compatibility.

Agreed. We generally try to communicate our breaking API changes in a
fair way amongst the known group of implementers, or at least make the
changes in such a way that they are not breaking (c.f., the expires_at
vs. expires_in) change.

I made a mistake in determining the impact that this change would have,
and for this I am sorry. We'll make sure to scrutinize with a more
careful eye in the future.

**Comment by pauldraper** [comment]
To any one searching the interwebs, the error in Bitbucket logs is

```
2018-08-08 18:11:37,265 ERROR [http-nio-7990-exec-7] paul @15CGODGx1091x321x1 i5r88x 10.0.2.107,10.0.2.14 "PUT /rest/git-lfs/storage/RIV/main/6f1986c018c8cc96864089f1204018c90252db157a8d89054b126456cdeec5af HTTP/1.1" c.a.b.i.s.g.l.http.HttpScmErrorUtils Un-mapped exception
javax.ws.rs.WebApplicationException: null
	at com.sun.jersey.server.impl.uri.rules.TerminatingRule.accept(TerminatingRule.java:66)
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
	at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
	at com.atlassian.analytics.client.filter.UniversalAnalyticsFilter.doFilter(UniversalAnalyticsFilter.java:92)
	at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:39)
	at com.atlassian.bitbucket.internal.xcode.web.XcodeUserAgentFilter.doFilter(XcodeUserAgentFilter.java:36)
```

I can't decide who's more at fault: Bitbucket for being unnecessarily sensitive, or Git LFS for adding a feature with nearly zero benefit. I guess mostly the former.

**Comment by pauldraper** [comment]
Bitbucket Server support tracked here: https://jira.atlassian.com/browse/BSERV-11047

**Comment by bturner** [comment]
> I can't decide who's more at fault: Bitbucket for being unnecessarily sensitive to Content-Type, or Git LFS for adding a feature with nearly zero benefit. I guess mostly the former.

Fascinating conclusion. Yes, it's Bitbucket Server's fault for implementing the LFS protocol exactly as documented and then failing when LFS was changed without warning (or even updating its own protocol documentation). Makes total sense.

Setting aside whose "fault" the breakage is, since it's a pointless conversation to have, we'll get Bitbucket Server's integration updated to handle the new normal. I'd encourage someone on the LFS side to update their documentation to clarify that upload content types are _not_ guaranteed to be "application/octet-stream" anymore, and that servers should be prepared for client-side content detection to mean varying content types can be used.

I'd also love to see something put in place that helps _broadcast_ these sorts of wire behavior changes so we can stay on top of them. Is there a mailing list, or some other source of truth I can subscribe to in order to be kept informed in the future?

**Comment by pauldraper** [comment]
The docs leave a lot to the imagination. Some values are constants and some are variable (e.g. Content-Type) and the docs don't say which is which.

(Though they do refer to "raw bytes", which I'm really not sure about...I think those are normal people bytes.)

**Comment by mmielke-ciena** [comment]
Thanks, @ttaylorr . I appreciate your response.

For the others, I think just expressions of frustrations and different levels of tongue-in-cheek. :-)

For future consideration, I want to mention these additional points:

The `lfs.contenttype` configuration as a work-around in generate is problematic because this isn't really a "per-user" setting. This is a "per-server" work-around. Let us just say that we have the 300+ users of Git LFS (a subset of our whole set) run the `git config` option. Now let's fast-forward to when Atlassian releases a patch to support Git LFS 2.5.0+. Now he have to get these same 300+ users to *undo* the configuration?

The `.lfsconfig` repository setting was mentioned as a possible solution to the above. However, it has a similar problem. We might have the repository on two different servers - and only one of these servers has the problem. Also, fast-forward to when Atlassian releases a patch to support Git LFS 2.5.0+ and now we have to do a commit to undo this change. We now have a historical record of Git LFS work arounds trapped in our product history!

I think the correct procedure for a change like this in the future would be:

1. Identify a community forum that has all major stakeholders. They may not know they need to be subscribed, so some amount of influence may be necessary to ensure a correct list of stakeholders are actively monitoring.
2. Pose the question as to what impact there would be to a wire change like this one. I would word it as "Who will get broken by this change?"
3. Obtain general agreement from the community that the wire change is the right change to do, and that all stake holders agree with it, for some definition of "agree" that is acceptable (majority rules?)
4. Negotiate a deployment schedule, that will provide reasonable opportunity for patches to be deployed. Many of these systems will require several weeks at minimum to deploy a change and run it through proper testing. Then, several weeks or more for customers to deploy the updates to their systems. I think 90 days is something close to the minimum for a serious product with millions of users, which Git LFS has become.

Obviously, much of the above can be avoided by not making breaking changes in the first place. Git LFS in its current form is effectively the reference implementation, and the documentation. When the specification isn't clear, but the code says "application/octet-stream", any reasonable reader is going to understand that the intent was for "application/octet-stream" to be used. Some might adhere to the "robustness principle of "Be conservative in what you do, be liberal in what you accept from others", but this knife cuts both ways. In this case, Git LFS could be conservative in what it does, and liberal in what it accepts from others. Without rules, we get lead down the road of implementation writers making up their own rules, and then expecting them to be followed as "de facto". So, I much prefer clear specifications and enforcement of these specifications, to "whatever works". :-)

For some of the above, I am probably preaching to the choir, and I'm only looking to record these notes for future reference. I think for the "git config" and ".lfsconfig" elements, though, that it might be new to people that these are not really good options for anybody with more than 1 user to support.

I think Git LFS growing pains. I do appreciate the recognition of the issue. Thank you.

**Comment by ttaylorr** [comment]
> Thanks, @ttaylorr. I appreciate your response.

You are most welcome :-).

> Let us just say that we have the 300+ users of Git LFS (a subset of
> our whole set) run the `git config` option.  Now let's fast-forward to
> when Atlassian releases a patch to support Git LFS 2.5.0+. Now he have
> to get these same 300+ users to *undo* the configuration?

I understand your frustration. Clearly this situation could have been
avoided if we decided not to merge the change that introduced
Content-Type detection.

For the sake of completeness, we did discuss a possibility of making
this transition somewhat easier (and I'm not sure that it's necessarily
required that _all_ users unset `lfs.contenttype` immediately after
Atlassian/GitLab change their implementation of the specification.)

> The `.lfsconfig` repository setting was mentioned as a possible
> solution to the above. However, it has a similar problem. We might
> have the repository on two different servers - and only one of these
> servers has the problem.

You can narrow the configuration by URL (as described -- I believe --
in `git-lfs-config(5)`), but I wanted more thoughts from the
@git-lfs/core team.

> I think the correct procedure for a change like this in the future
> would be:
>
> 1. Identify a community forum that has all major stakeholders. They
> may not know they need to be subscribed, so some amount of influence
> may be necessary to ensure a correct list of stakeholders are actively
> monitoring.

We already have a `@git-lfs/releases` team, which are humans interested
in releases of Git LFS (i.e., the Git for Windows maintainer is a member
because they bundle Git into their distribution of it). A
`@git-lfs/implementers` team might be appropriate, too.

That said, I'm not sure that some of these processes below fit into the
development model that Git LFS has today. The vast majority of our
changes are non-breaking (this is the only one that I can remember of
more than two years of working on Git LFS full-time).

So, some of the more heavy-weight items like (4) might be difficult to
implement given the current nature of things. That said, I am more than
thrilled to notify a future @git-lfs/implementers team that could host
representatives from other hosting companies and people who might have a
stake in the issue.

That all said, I am out of the office starting tomorrow (9 August, 2018)
and will be returning in two weeks (on 20 August, 2018), so the creation
of that team and inviting members will have to wait until then.

My thoughts in one sentence are mostly, "let's create a new team, with
representation from all of the parties that might be interested in a
breaking API change and notify them reasonably before shipping a release
that breaks the API".

Does that sound OK?

**Comment by mmielke-ciena** [comment]
Yes, please.

Side note:

For .lfsconfig I neglected my killer argument. :-) In our Bitbucket instance we have 12,479 repositories. Some are forks, and many don't use Git LFS, but the point here is that configuring .lfsconfig in the repository means changing some percentage of the 12,479 repositories to disable lfs.contenttype. It would probably be easier to target the 300+ users. :-)

**Comment by benhumphreys** [comment]
For Bitbucket Server we have implemented a hotfix that can be applied to all affected versions, from 4.3.0 (where LFS support was added) through to the current 5.13.0 release. Details, including the hotfix plugin can be found on issue [BSERV-11047](https://jira.atlassian.com/browse/BSERV-11047?focusedCommentId=1853063&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1853063).

The hotfix is delivered as a plugin that implements a simple Servlet Filter that overrides the Content-Type for all uploads to the embedded Git LFS object store.

A patch release of Bitbucket Server will be available in the near future with official support for Git LFS client versions 2.5.0+.

**Comment by ttaylorr** [comment]
> A patch release of Bitbucket Server will be available in the near
> future with official support for Git LFS client versions 2.5.0+.

Thank you very much. We'll make sure that changes like these go more
smoothly in the future.

**Comment by ttaylorr** [comment]
> Yes, please.

Most welcome. I created the @git-lfs/implementers team and invited
@benhumphreys and @bturner to start.

**Comment by gtmaskall** [comment]
Well, this consumed a few hours of my time!
tl;dr - On a linux system, git-lfs suddenly started boaking when pushing LFS objects to Atlassian's bitbucket. All I got was a rather unhelpful, cryptic

> Fatal error: null
> Fatal error: null
> error: failed to push some refs to 'https://....

This issue was helpful, and the
`git config lfs.contenttype false`
config setting resolved it.

I prepended GIT_TRACE=1 GIT_CURL_VERBOSE=1 to my git push command and captured stderr before and after the above git config change. Using meld, I homed in on the key difference, which hopefully I summarise effectively below.

Before the config change, when the push fails, in response to an HTTP: PUT I get:

> > Content-Length: 25965
> > Content-Type: application/x-gzip
> > User-Agent: git-lfs/2.5.1 (GitHub; linux amd64; go 1.10.3)
> > 
> 11:35:50.058866 trace git-lfs: HTTP: 500
> 
> 
> < HTTP/1.1 500 Internal Server Error
> < Connection: close
> < Transfer-Encoding: chunked
> < Content-Type: application/vnd.git-lfs+json
> < Date: Thu, 23 Aug 2018 10:35:49 GMT
> < Server: Apache-Coyote/1.1
> < Set-Cookie: JSESSIONID=68B779320B6E2CEF96E5B8FF0A7FF32A; Path=/; Secure; HttpOnly
> < X-Arequestid: @RZS2QWx275x705243x2
> < X-Asen: SEN-5992902
> < X-Asessionid: yqnqir
> < X-Auserid: 1758
> < X-Ausername: notmyname
> < X-Content-Type-Options: nosniff
> < 
> 11:35:50.059060 trace git-lfs: HTTP: {"message":"null"}
> {"message":"null"}11:35:50.059256 trace git-lfs: tq: retrying object b32b2de8e0344a22e82ee6823ce2eefc3e1285258348d60461736e6dbe47ccac: Fatal error: null
> 11:35:50.059297 trace git-lfs: tq: enqueue retry #1 for "b32b2de8e0344a22e82ee6823ce2eefc3e1285258348d60461736e6dbe47ccac" (size: 51683)
> 11:35:50.187253 trace git-lfs: HTTP: 500

(It clearly retries and fails again, but that's the key bit)
Whereas after making the git config change, I get:

> > Content-Length: 25965
> > User-Agent: git-lfs/2.5.1 (GitHub; linux amd64; go 1.10.3)
> > 
> 11:46:48.739612 trace git-lfs: HTTP: 200

Notice no Content-Type: line and then an HTTP: 200 instead of 500.

**Comment by ttaylorr** [comment]
> Notice no Content-Type: line and then an HTTP: 200 instead of 500.

Interesting! It looks like there is certainly a bug with the
`lfs.[url].contentType` setting, in that when it is 'on' we'll send the
detected content type of the HTTP body, but when it is 'off', we'll send
nothing at all.

I'm surprised to learn that the former causes a problem, but the later
does not.

In any case, @bk2204 has graciously fixed this issue in
https://github.com/git-lfs/git-lfs/pull/3201.

**Comment by triplef** [comment]
> > Fatal error: null
> > Fatal error: null
> > error: failed to push some refs to 'https://....
> 
> This issue was helpful, and the
> `git config lfs.contenttype false`
> config setting resolved it.

I had the exact same issue using the [Tower Git](https://www.git-tower.com) client with Bitbucket Server v5.9.0, and setting `lfs.contenttype` resolved it for me as well. Thank you for documenting this workaround @gtmaskall.

**Comment by imuneer** [comment]
> 
> After updating my git-lfs package to 2.5.0, when I try to push lfs files to the server the upload bitrate and progress go a bit nuts,
> Oh, no -- I'm sorry that there was a regression and I'm sure that we'll be able to get to the bottom of whatever is going on.
> and eventually I get the following error: Fatal error: Unable to parse HTTP response for PUT https://<servername>/rest/git-lfs/storage/<path to lfs object>: unexpected EOF error: failed to push some refs to 'ssh://git@<project>.git'
> Can you reproduce the issue reliably? If so, please post the output (with your sensitive credentials obfuscated) by re-running the command with the following in your environment: `GIT_TRACE=1` and `GIT_CURL_VERBOSE=1`. That should help us pinpoint a little closer what the issue might be. Another thing that you could try is: $ git config lfs.contenttype 0 Which will disable "Content-Type:" HTTP header inference.

`git config lfs.contenttype 0`

I was having the same issue, This helped me as well.
